server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -2454,6 +2520,19 @@ public static IndexMetadata fromXContent(XContentParser parser, Map<String, Mapp
                         case KEY_STATS:
                             builder.stats(IndexMetadataStats.fromXContent(parser));
                             break;
+                        case KEY_FIELDS_FOR_MODELS:
+                            // TODO: Could probably make this more efficient
+                            Map<String, Set<String>> fieldsForModels = parser.map(HashMap::new, XContentParser::list)
+                                .entrySet()
+                                .stream()
+                                .collect(
+                                    Collectors.toMap(
+                                        Map.Entry::getKey,
+                                        v -> v.getValue().stream().map(Object::toString).collect(Collectors.toUnmodifiableSet())
```

Comment:
I wonder if it is necessary to use `unModifiableSet` if a `builder.putMapping` below will create a unmodifiable set again?

Reply:
You're right, we don't need to create an unmodifiable set here because `builder.fieldsForModels` will handle making everything immutable.

However, is there any efficiency gain in using `.collect(Collectors.toSet())` instead? It would be better if we could deserialize directly to a set (immutable or not). Are you aware of any XContent parsers that do this?